src/core/symbology/qgslinearreferencingsymbollayer.cpp

```diff
@@ -0,0 +1,1003 @@
+    }
+
+    const double angleRadians = ( mRotateLabels ? angle : 0 ) * M_PI / 180.0;
+    const double dx = labelOffsetPainterUnits.x() * std::sin( angleRadians + M_PI_2 )
+    + labelOffsetPainterUnits.y() * std::sin( angleRadians );
+    const double dy = labelOffsetPainterUnits.x() * std::cos( angleRadians + M_PI_2 )
+    + labelOffsetPainterUnits.y() * std::cos( angleRadians );
+
+    QgsTextRenderer::drawText( QPointF( pt.x() + dx, pt.y() + dy ), angleRadians, Qgis::TextHorizontalAlignment::Left, { mNumericFormat->formatDouble( labelValue, numericContext ) }, context.renderContext(), mTextFormat );
+
+    return true;
+  } );
+}
+
+void QgsLinearReferencingSymbolLayer::renderPolylineVertex( const QgsLineString *line, QgsSymbolRenderContext &context, double skipMultiples, const QPointF &labelOffsetPainterUnits, double averageAngleLengthPainterUnits, bool showMarker )
+{
+  // let's simplify the logic by ALWAYS using the averaging approach for angles, and just
+  // use a very small distance if the user actually set this to 0. It'll be identical
+  // results anyway...
+  averageAngleLengthPainterUnits = std::max( averageAngleLengthPainterUnits, 0.1 );
+
+  QgsNumericFormatContext numericContext;
+  numericContext.setExpressionContext( context.renderContext().expressionContext() );
+
+  const double *xData = line->xData();
+  const double *yData = line->yData();
+  const double *zData = line->is3D() ? line->zData() : nullptr;
+  const double *mData = line->isMeasure() ? line->mData() : nullptr;
+  const int size = line->numPoints();
+  if ( size < 2 )
+    return;
+
+  std::unique_ptr< QgsLineString > painterUnitsGeometry( line->clone() );
+  if ( context.renderContext().coordinateTransform().isValid() )
+  {
+    painterUnitsGeometry->transform( context.renderContext().coordinateTransform() );
+  }
+  painterUnitsGeometry->transform( context.renderContext().mapToPixel().transform() );
+  const double *xPainterUnits = painterUnitsGeometry->xData();
+  const double *yPainterUnits = painterUnitsGeometry->yData();
+
+  double currentDistance = 0;
+  double prevX = *xData;
+  double prevY = *yData;
+
+  for ( int i = 0; i < size; ++i )
+  {
+    if ( context.renderContext().renderingStopped() )
+      break;
+
+    double thisX = *xData++;
+    double thisY = *yData++;
+    double thisZ = zData ? *zData++ : 0;
+    double thisM = mData ? *mData++ : 0;
+    double thisXPainterUnits = *xPainterUnits++;
+    double thisYPainterUnits = *yPainterUnits++;
+
+    const double thisSegmentLength = QgsGeometryUtilsBase::distance2D( prevX, prevY, thisX, thisY );
+    currentDistance += thisSegmentLength;
+
+    if ( skipMultiples > 0 && qgsDoubleNear( std::fmod( currentDistance,  skipMultiples ), 0 ) )
+    {
+      prevX = thisX;
+      prevY = thisY;
+      continue;
+    }
+
+    const QPointF pt = pointToPainter( context, thisX, thisY, thisZ );
+
+    double calculatedAngle = 0;
```

Comment:
Can we move this down to the line that starts with calculatedAngle = std::fmod( QgsGeometryUtilsBase ... ? It's not used before then, it'd make the code reading easier.

Reply:
Done